File: ProblemSolve/923.cpp
/**************************
 * 3Sum With Multiplicity *
 **************************/

class Solution {
public:
    int threeSumMulti(vector<int>& arr, int target) {
        unordered_map<int, int> hash;
        int ans = 0;
        int mod = 1e9 + 7;
        
        for(int i=0; i<arr.size(); ++i){
            ans = (ans + hash[target - arr[i]]) % mod;
            for(int j=0; j<i; ++j){
                hash[arr[i] + arr[j]]++;
            }
        }
        
        return ans;
    }
};

File: ProblemSolve/923.cs
/**************************
 * 3Sum With Multiplicity *
 **************************/

public class Solution {
    public int ThreeSumMulti(int[] arr, int target) {
        Dictionary<int, int> hash = new Dictionary<int, int>();
        
        int ans = 0, mod = (int)1e9 + 7;
        
        for(int i=0; i<arr.Length; ++i){
            if(hash.TryGetValue(target - arr[i], out int val))
                ans = (ans + val) % mod;
            
            for(int j=0; j<i; ++j){
                if(!hash.TryGetValue(arr[i] + arr[j], out val))
                    hash[arr[i] + arr[j]] = 1;
                else
                    hash[arr[i] + arr[j]]++;
            }
        }
        
        return ans;
    }
}